File: android/moqkit/moqkit/src/main/java/com/swmansion/moqkit/publish/MediaTrackOutput.kt
package com.swmansion.moqkit.publish

import dev.moq.BroadcastProducer
import dev.moq.Frame
import dev.moq.Init
import dev.moq.MediaProducer

/** One encoder generation; late callbacks cannot recreate a finished media track. */
internal class MediaTrackOutput(
    private val broadcast: BroadcastProducer,
    private val format: String,
    private val onActive: () -> Unit,
    private val onError: (Exception) -> Unit,
) {
    private var closed = false
    private var producer: MediaProducer? = null

    @Synchronized
    fun write(data: ByteArray, initData: ByteArray?, timestampUs: Long) {
        if (closed) return
        try {
            if (producer == null) {
                if (initData == null) return
                producer = broadcast.publishMedia(Init(format = format, data = initData, video = null))
                onActive()
            }
            producer?.writeFrame(Frame(payload = data, timestampUs = timestampUs.toULong()))
        } catch (e: Exception) {
            fail(e)
        }
    }

    @Synchronized
    fun fail(error: Exception) {
        if (closed) return
        finish()
        onError(error)
    }

    @Synchronized
    fun finish() {
        closed = true
        try { producer?.finish() } catch (_: Exception) {}
        try { producer?.close() } catch (_: Exception) {}
        producer = null
    }
}

File: ios/Sources/MoQKit/Publish/MediaTrackOutput.swift
import Foundation
import Moq

/// One encoder generation. Closing it fences late output before removing its catalog entry.
final class MediaTrackOutput {
    private let lock = NSLock()
    private var closed = false
    private var producer: Moq.MediaProducer?
    private let broadcast: Moq.BroadcastProducer
    private let format: String
    private let onActive: () -> Void
    private let onError: (Error) -> Void

    init(broadcast: Moq.BroadcastProducer, format: String,
         onActive: @escaping () -> Void, onError: @escaping (Error) -> Void) {
        self.broadcast = broadcast
        self.format = format
        self.onActive = onActive
        self.onError = onError
    }

    func write(_ data: Data, initData: Data?, timestampUs: UInt64) {
        lock.lock()
        defer { lock.unlock() }
        guard !closed else { return }
        do {
            if producer == nil {
                guard let initData else { return }
                producer = try broadcast.publishMedia(format: format, initData: initData)
                onActive()
            }
            try producer?.writeFrame(data, timestampUs: timestampUs)
        } catch {
            closed = true
            try? producer?.finish()
            producer = nil
            onError(error)
        }
    }

    func fail(_ error: Error) {
        lock.lock()
        defer { lock.unlock() }
        guard !closed else { return }
        closed = true
        try? producer?.finish()
        producer = nil
        onError(error)
    }

    func finish() {
        lock.lock()
        defer { lock.unlock() }
        closed = true
        try? producer?.finish()
        producer = nil
    }
}
